README.md
# Flutter Notification Permissions
Package to check for and ask for Notification Permissions on iOS and Android.

## Checking Notification Permission Status
```dart
Future<PermissionStatus> permissionStatus =
    NotificationPermissions.getNotificationPermissionStatus();
```

This method will return an enum with the following values:

```dart
enum PermissionStatus {
  provisional, // iOS Only
	granted,
	unknown,
	denied
}
```

In iOS, a permission is `unknown` when the user hasn’t accepted or refuse the notification permissions. In Android this state will never occur, since the permission will be `granted` by default and it will be `denied` if the user goes to the app settings and turns off notifications for the app. The `provisional` status will also offer the same behavior.

## Requesting Notification Permissions
If the `PermissionStatus` is `denied` or `unknown`, we can ask the user for the Permissions:
```dart
Future<PermissionStatus> permissionStatus = NotificationPermissions.requestNotificationPermissions({NotificationSettingsIos iosSettings, bool openSettings});
```

On Android, if the permission is `denied`, this method will open the app settings.

In iOS, if the permission is `unknown` or `provisional`, it will show an alert window asking the user for the permission. On the other hand, if the permission is `denied` it has the same behaviour as Android, opening the app settings.
Also in iOS if you set `openSettings` to false settings window won't be opened. You will get `denied` status. 
`NotificationPermissions.requestNotificationPermissions` returns status after user select answer from native permission popup.

Note: if the permission is `granted`, this method will not do anything.

## iOS Error: Swift.h not found

If your project is in Objective-C, you will have to do the changes referenced in [this SO post](https://stackoverflow.com/a/53453243/4499889) in order to solve the issue:

```
'notification_permissions/notification_permissions-Swift.h' file not found
    #import <notification_permissions/notification_permissions-Swift.h>
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    1 error generated.
```

Add `use_frameworks!` in target Runner in your `(YOUR_PROJECT)/ios/Podfile`

```
target 'Runner' do
  use_frameworks! # Add here
```

## Special Thanks
Special thanks to [fedecastelli](https://github.com/fedecastelli) for helping me in the Swift Code!


CHANGELOG.md
# 0.6.1

* Suppresses Android deprecation warning [nohli]

# 0.6.0

* Supports Android V2 embedding [ened]

# 0.5.0

* Adds null safety [anoop4real]

# 0.4.8

* Fixes iOS Privisional Status not opening [#30](https://github.com/Vanethos/flutter_notification_permissions/pull/30), 
[DuncanKelbie]

# 0.4.7

* Adds iOS Provisional Status [#28](https://github.com/Vanethos/flutter_notification_permissions/pull/28), [DuncanKelbie]

# 0.4.6
* Fixes open android Settings via [#25](https://github.com/Vanethos/flutter_notification_permissions/pull/25) , [britannio]

# 0.4.5
* Fixes Android giving unknown permission error [Zazo032]

# 0.4.4
* Corrected the path to the license file in .podspec
* Define CLANG module, so that the plugin can be used statically
* Ensure correct version is listed in .podspec

# 0.4.3 
* Added option to open settings or not in iOS when permition is denied [pawlowskim]
* Adding possibility for iOS 10.0 to wait for call requestNotificationPermissions (iOS permission dialog) [kuzmisin]
* Fix missing result() call for Android when PERMISSION_DENIED [kuzmisin]
* Fixed warnings on XCode [ened]

# 0.4.0 
* **Breaking change** Changed implementation to use static methods. Changed `IosNotificationSettings` to `NotificationSettingsIos` to avoid conflict with Firebase

# 0.3.1
* Changed the README to present the quick-fix for installing this plugin in a Objective-C Flutter Project

# 0.3.0
* Replace Kotlin implementation with simple Java to ease future maintenance.

## 0.2.2
* Added support for latest kotlin version

## 0.2.1
* Port iOS codebase to Swift 4.2

## 0.2.0
* **Breaking change**. Migrate from the deprecated original Android Support Library to AndroidX. This shouldn't result in any functional changes, but it requires any Android apps using this plugin to [also migrate](https://developer.android.com/jetpack/androidx/migrate) if they're using the original support library.

## 0.1.0
* Initial release of the package


ios/Classes/SwiftNotificationPermissionsPlugin.swift
import Flutter
import UIKit
import UserNotifications

public class SwiftNotificationPermissionsPlugin: NSObject, FlutterPlugin {
  var permissionGranted:String = "granted"
  var permissionUnknown:String = "unknown"
  var permissionDenied:String = "denied"
  var permissionProvisional:String = "provisional"

  public static func register(with registrar: FlutterPluginRegistrar) {
      let channel = FlutterMethodChannel(name: "notification_permissions", binaryMessenger: registrar.messenger())
      let instance = SwiftNotificationPermissionsPlugin()
      registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      if (call.method == "requestNotificationPermissions") {
          // check if we can ask for permissions
          getNotificationStatus(completion: { status in
              if (status == self.permissionUnknown || status == self.permissionProvisional) {
				  if #available(iOS 10.0, *) {
					  let center = UNUserNotificationCenter.current()
					  var options = UNAuthorizationOptions()
					  if let arguments = call.arguments as? Dictionary<String, Bool> {
						  if(arguments["sound"] != nil){
							options.insert(.sound)
						  }
						  if(arguments["alert"] != nil){
							options.insert(.alert)
						  }
						  if(arguments["badge"] != nil){
							options.insert(.badge)
						  }
					  }
					  center.requestAuthorization(options: options) { (success, error) in
						  if error == nil {
							  if success == true {
								  result(self.permissionGranted)
							  }
							  else {
								  result(self.permissionDenied)
							  }
						  }
						  else {
							  result(error?.localizedDescription)
						  }
					}
				  } else {
				      // If the permission status is unknown, the user hasn't denied it
					  var notificationTypes = UIUserNotificationType(rawValue: 0)
					  if let arguments = call.arguments as? Dictionary<String, Bool> {
						  if arguments["sound"] != nil {
							  notificationTypes.insert(UIUserNotificationType.sound)
						  }
						  if arguments["alert"] != nil  {
							  notificationTypes.insert(UIUserNotificationType.alert)
						  }
						  if arguments["badge"] != nil  {
							  notificationTypes.insert(UIUserNotificationType.badge)
						  }
						  var settings: UIUserNotificationSettings? = nil

						  settings = UIUserNotificationSettings(types: notificationTypes, categories: nil)

						  if let settings = settings {
							  UIApplication.shared.registerUserNotificationSettings(settings)
						  }
						  self.getNotificationStatus(completion: { status in
						  	result(status)
						  });
					  }
				  }
              } else if (status == self.permissionDenied) {
                  // The user has denied the permission he must go to the settings screen
                  if let arguments = call.arguments as? Dictionary<String, Bool> {
					  if (arguments["openSettings"] != nil && arguments["openSettings"] == false)  {
						  result(self.permissionDenied)
						  return
					  }
				  }
                  if let url = URL(string:UIApplication.openSettingsURLString) {
                      if UIApplication.shared.canOpenURL(url) {
                          if #available(iOS 10.0, *) {
                              UIApplication.shared.open(url, options: convertToUIApplicationOpenExternalURLOptionsKeyDictionary([:]), completionHandler: nil)
                          } else {
                              UIApplication.shared.openURL(url)
                          }
                      }
                  }
                  result(nil)
              } else {
                  result(nil)
              }
          })
      } else if (call.method == "getNotificationPermissionStatus") {
          getNotificationStatus(completion: { status in
              result(status)
          })
      } else {
          result(FlutterMethodNotImplemented)
      }
  }

  func getNotificationStatus(completion: @escaping ((String) -> Void)) {
      if #available(iOS 10.0, *) {
          let current = UNUserNotificationCenter.current()
          current.getNotificationSettings(completionHandler: { settings in
              if settings.authorizationStatus == .notDetermined {
                  completion(self.permissionUnknown)
              } else if settings.authorizationStatus == .denied {
                  completion(self.permissionDenied)
              } else if settings.authorizationStatus == .authorized {
                  completion(self.permissionGranted)
              } else if #available(iOS 12.0, *){
                  if (settings.authorizationStatus == .provisional){
                      completion(self.permissionProvisional)
                  }
              }
          })
      } else {
          // Fallback on earlier versions
          if UIApplication.shared.isRegisteredForRemoteNotifications {
              completion(self.permissionGranted)
          } else {
              completion(self.permissionDenied)
          }
      }
  }
}

// Helper function inserted by Swift 4.2 migrator.
fileprivate func convertToUIApplicationOpenExternalURLOptionsKeyDictionary(_ input: [String: Any]) -> [UIApplication.OpenExternalURLOptionsKey: Any] {
	return Dictionary(uniqueKeysWithValues: input.map { key, value in (UIApplication.OpenExternalURLOptionsKey(rawValue: key), value)})
}
